docs(gpu): mig with k8s (#2165)
scaleway/docs-content: compute/gpu/how-to/use-mig-with-kubernetes.mdx

                        nvidia.com/gpu.engines.encoder=0
                        nvidia.com/gpu.engines.jpeg=1
                        nvidia.com/gpu.engines.ofa=0
                        nvidia.com/gpu.family=hopper
                        nvidia.com/gpu.machine=SCW-H100-1-80G
                        nvidia.com/gpu.memory=9856
                        nvidia.com/gpu.multiprocessors=14
                        nvidia.com/gpu.present=true
                        nvidia.com/gpu.product=NVIDIA-H100-PCIe-MIG-1g.10gb
                        nvidia.com/gpu.replicas=1
                        nvidia.com/gpu.slices.ci=1
                        nvidia.com/gpu.slices.gi=1
                        nvidia.com/mig.capable=true
                        nvidia.com/mig.config=all-1g.10gb
                        nvidia.com/mig.config.state=success
                        nvidia.com/mig.strategy=single
                        nvidia.com/gpu-driver-upgrade-enabled: true
    nvidia.com/gpu:     7
    nvidia.com/gpu:     7
    nvidia.com/gpu     0           0
    ```

## Deploy containers that use NVIDIA MIG technology partitions

1. Write a deployment file to deploy 8 pods executing Nvidia SMI. 

>     Open a text editor of your choice and create a deployment file `deploy-mig.yaml`, then paste the following content into the file, save it and exit the editor:

Author: No it is part of the first step

---
meta:
  title: How to use NVIDIA MIG technology with Kubernetes
  description: This section provides information about MIG with Kubernetes
content:
  h1: How to use NVIDIA MIG technology with Kubernetes
  paragraph: This section provides information about MIG with Kubernetes
tags: NVIDIA H100 MIG multi-instance gpu
dates:
  validation: 2023-09-19
  posted: 2023-09-19
categories:
  - compute
---

<Message type="important">
  * Scaleway offers GPU Instances such as H100 PCIe GPU Instances, RENDER-S GPU Instance, etc.
  * Nvidia uses the term *"GPU instance"* to designate a MIG partition of a GPU (MIG= Multi-Instance GPU)
  * To avoid confusion, we will use the term GPU Instance in this document to refer to the Scaleway GPU Instance, and *"MIG partition"* in the context of the MIG feature.
</Message>

NVIDIA Multi-Instance GPU (MIG) is a powerful feature that allows you to divide a single NVIDIA GPU into multiple smaller partitions, each with its dedicated GPU resources, such as memory and compute units.
This technology is particularly valuable in Kubernetes (K8s) environments, where efficient resource allocation is crucial for running diverse workloads efficiently.

> In this guide, we will explore capabilities of NVIDIA MIG within a Kubernetes cluster. We cover the steps required to set up and configure MIG-enabled GPUs to use with Kubernetes, to maximize GPU usage and ensure workload isolation and performance predictability.

Reviewer: ```suggestion
In this guide, we will explore the capabilities of NVIDIA MIG within a Kubernetes cluster. We cover the steps required to set up and configure MIG-enabled GPUs to use with Kubernetes, to maximize GPU usage and ensure workload isolation and performance predictability.
```

1. Reconfigure the GPU into seven MIG `1g.10gb` partitions by overwriting the existing labels:
    ```
    % kubectl label nodes scw-k8s-jovial-dubinsky-pool-h100-93a072191d38 nvidia.com/mig.config=all-1g.10gb --overwrite 
    node/scw-k8s-jovial-dubinsky-pool-h100-93a072191d38 labeled
    ```

2. Check the status of Nvidia SMI in the Nvidia driver container:
    ```
    % kubectl exec nvidia-driver-daemonset-8t89m  -t -n kube-system -- nvidia-smi -L
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-222504cc-4a15-589b-8ec8-dbc02e6fb378)
    MIG 1g.10gb     Device  1: (UUID: MIG-fdfd2afa-5cbd-5d1d-b1ae-6f0e13cc0ff8)
    MIG 1g.10gb     Device  2: (UUID: MIG-b2925bc6-41ca-5ccd-bf5e-24259386f88e)
    MIG 1g.10gb     Device  3: (UUID: MIG-083c76fc-5d21-5322-9d50-c8e21a01852f)
    MIG 1g.10gb     Device  4: (UUID: MIG-13d8a181-5bc1-5527-9a0f-9c3f9cc1d89e)
    MIG 1g.10gb     Device  5: (UUID: MIG-db99bb81-dde3-5c95-9778-daa291fce210)
    MIG 1g.10gb     Device  6: (UUID: MIG-2d636152-57c2-5e73-9654-b1d21d6d21fb)
    ```
    Seven MIG `1g.10gb` partitions are available now.

> 3. Have a look the Nvidia labels on the node (note the label `nvidia.com/mig.config=all-1g.10gb` and `nvidia.com/gpu.product=NVIDIA-H100-PCIe-MIG-1g.10gb`):

Reviewer: ```suggestion
3. Look at the Nvidia labels on the node (note the label `nvidia.com/mig.config=all-1g.10gb` and `nvidia.com/gpu.product=NVIDIA-H100-PCIe-MIG-1g.10gb`):
```

    apiVersion: v1
    kind: Pod
    metadata:
    name: test-6
    spec:
    restartPolicy: OnFailure
    containers:
    - name: gpu-test
        image: nvcr.io/nvidia/pytorch:23.07-py3
        command: [ "nvidia-smi" ]
        args: ["-L" ]
        resources:
        limits:
            nvidia.com/gpu: 1
    nodeSelector:
        nvidia.com/gpu.product : NVIDIA-H100-PCIe-MIG-1g.10gb

    ---

    apiVersion: v1
    kind: Pod
    metadata:
    name: test-7
    spec:
    restartPolicy: OnFailure
    containers:
    - name: gpu-test
        image: nvcr.io/nvidia/pytorch:23.07-py3
        command: [ "nvidia-smi" ]
        args: ["-L" ]
        resources:
        limits:
            nvidia.com/gpu: 1
    nodeSelector:
        nvidia.com/gpu.product : NVIDIA-H100-PCIe-MIG-1g.10gb

    ---

    apiVersion: v1
    kind: Pod
    metadata:
    name: test-8
    spec:
    restartPolicy: OnFailure
    containers:
    - name: gpu-test
        image: nvcr.io/nvidia/pytorch:23.07-py3
        command: [ "nvidia-smi" ]
        args: ["-L" ]
        resources:
        limits:
            nvidia.com/gpu: 1
    nodeSelector:
        nvidia.com/gpu.product : NVIDIA-H100-PCIe-MIG-1g.10gb
    ```

> 

Reviewer: ```suggestion
```

2. Deploy the pods:
    ```
    % kubectl create -f deploy-mig.yaml
    pod/test-1 created
    pod/test-2 created
    pod/test-3 created
    pod/test-4 created
    pod/test-5 created
    pod/test-6 created
    pod/test-7 created
    pod/test-8 created
    ```

> 3. Display the logs of the pods. The pods are printing their UUID with the `nvidia-smi` command:

Reviewer: ```suggestion
3. Display the logs of the pods. The pods print their UUID with the `nvidia-smi` command:
```

    MIG 1g.10gb     Device  0: (UUID: MIG-2d636152-57c2-5e73-9654-b1d21d6d21fb)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-db99bb81-dde3-5c95-9778-daa291fce210)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-13d8a181-5bc1-5527-9a0f-9c3f9cc1d89e)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-b2925bc6-41ca-5ccd-bf5e-24259386f88e)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-fdfd2afa-5cbd-5d1d-b1ae-6f0e13cc0ff8)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-222504cc-4a15-589b-8ec8-dbc02e6fb378)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-083c76fc-5d21-5322-9d50-c8e21a01852f)
    GPU 0: NVIDIA H100 PCIe (UUID: GPU-717ef73c-2d43-4fdc-76d2-1cddef4863bb)
    MIG 1g.10gb     Device  0: (UUID: MIG-fdfd2afa-5cbd-5d1d-b1ae-6f0e13cc0ff8)
    ```

>     As you can see, seven pods have been executed on different MIG partitions while the eigth pod had to wait for one of the seven MIG partitions to become available in order to be executed.

Reviewer: ```suggestion
    As you can see, seven pods have been executed on different MIG partitions, while the eighth pod had to wait for one of the seven MIG partitions to become available in order to be executed.
```
